src/instrument.cpp

```diff
@@ -919,43 +973,70 @@ Result BytecodeRewriter::rewriteClass() {
     put(cpool_start, cpool_end - cpool_start);
 
     putConstant(JVM_CONSTANT_Methodref, _cpool_len + 1, _cpool_len + 2);
+    _recordEntry_cpool_idx = _cpool_len;
     putConstant(JVM_CONSTANT_Class, _cpool_len + 3);
     putConstant(JVM_CONSTANT_NameAndType, _cpool_len + 4, _cpool_len + 5);
     putConstant("one/profiler/Instrument");
     putConstant("recordEntry");
     putConstant("()V");
 
     putConstant(JVM_CONSTANT_Methodref, _cpool_len + 1, _cpool_len + 7);
+    _recordExit_cpool_idx = _cpool_len + 6;
     putConstant(JVM_CONSTANT_NameAndType, _cpool_len + 8, _cpool_len + 9);
     putConstant("recordExit");
-    putConstant(_latency > 0 ? "(JJ)V" : "(J)V");
+    putConstant("(JJ)V");
+
+    putConstant(JVM_CONSTANT_Methodref, _cpool_len + 1, _cpool_len + 11);
+    _recordExit_latency0_cpool_idx = _cpool_len + 10;
+    putConstant(JVM_CONSTANT_NameAndType, _cpool_len + 8, _cpool_len + 12);
+    putConstant("(J)V");
 
-    putConstant(JVM_CONSTANT_Methodref, _cpool_len + 11, _cpool_len + 12);
-    putConstant(JVM_CONSTANT_Class, _cpool_len + 13);
-    putConstant(JVM_CONSTANT_NameAndType, _cpool_len + 14, _cpool_len + 15);
+    putConstant(JVM_CONSTANT_Methodref, _cpool_len + 14, _cpool_len + 15);
+    _nanoTime_cpool_idx = _cpool_len + 13;
+    putConstant(JVM_CONSTANT_Class, _cpool_len + 16);
+    putConstant(JVM_CONSTANT_NameAndType, _cpool_len + 17, _cpool_len + 18);
     putConstant("java/lang/System");
     putConstant("nanoTime");
     putConstant("()J");
 
-    putLongConstant(_latency >= 0 ? (u64) _latency : 0);
-
+    // Flushed later to the buffer, after latency-related constants are written to the cpool
     u16 access_flags = get16();
-    put16(access_flags);
-
     u16 this_class = get16();
-    put16(this_class);
 
     u16 class_name_index = _cpool[this_class]->info();
     _class_name = _cpool[class_name_index];
-    if (!_class_name->matches(_target_class, _target_class_len)) {
-        return Result::CLASS_DOES_NOT_MATCH;
-    }
     // We should not instrument classes from one.profiler.* to avoid infinite recursion
     if (_class_name->info() >= PROFILER_PACKAGE_LEN &&
-        strncmp(_class_name->utf8(), PROFILER_PACKAGE, PROFILER_PACKAGE_LEN) == 0) {
+    strncmp(_class_name->utf8(), PROFILER_PACKAGE, PROFILER_PACKAGE_LEN) == 0) {
         return Result::PROFILER_CLASS;
     }
 
+    for (const auto& target : *_targets) {
```

Comment:
This is kind of redundant, redoing the search is the alternative to taking a `const MethodTargets*` in the constructor. Let me know which one you prefer @apangin 

We cannot remove the logic from `ClassFileLoadHook` because it would be inefficient

Reply:
99bc73433e0fb25d09843fdfcbfd78f1482a5903